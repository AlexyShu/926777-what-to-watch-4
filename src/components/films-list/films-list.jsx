import React, {PureComponent} from "react";
import {connect} from "react-redux";
import PropTypes from "prop-types";
import FilmCard from "../film-card/film-card.jsx";

const TIMEOUT = 1000;

class FilmsList extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      selectedFilmId: null,
      isPlaying: false
    };

    this.movieCardMouseOverHandler = this.movieCardMouseOverHandler.bind(this);
    this.movieCardMouseOutHandler = this.movieCardMouseOutHandler.bind(this);
    this.togglePlay = this.togglePlay.bind(this);
  }

  togglePlay(selectedFilmId) {
    setTimeout(() => {
      if (this.state.selectedFilmId === selectedFilmId) {
        this.setState((prevState) => ({
          isPlaying: !prevState.isPlaying
        }));
      }
    }, TIMEOUT);
  }

  movieCardMouseOverHandler(selectedFilmId) {
    this.setState(
        () => ({
          selectedFilmId
        }),
        () => this.togglePlay(selectedFilmId)
    );
  }

  movieCardMouseOutHandler() {
    this.setState(() => ({
      selectedFilmId: null,
      isPlaying: false
    }));
  }

  render() {
    const {films, onFilmCardClick, filmsCount} = this.props;
    return <div className="catalog__movies-list">
      {films.map((film, i) =>
        <FilmCard
          key = {film.name + i}
          film = {film}
          onFilmCardClick = {onFilmCardClick}
          onMovieCardMouseOver={() => this.movieCardMouseOverHandler(i)}
          onMovieCardMouseOut={this.movieCardMouseOutHandler}
          isPlaying={
            this.state.selectedFilmId === i && this.state.isPlaying
          }
        />
      ).slice(0, filmsCount)}
    </div>;
  }
}

const mapStateToProps = (state) => ({
  films: state.films
});

FilmsList.propTypes = {
  films: PropTypes.arrayOf(
      PropTypes.shape({
        name: PropTypes.string.isRequired,
        posterSrc: PropTypes.string.isRequired,
        genre: PropTypes.string.isRequired,
        year: PropTypes.number.isRequired,
        ratingScore: PropTypes.string.isRequired,
        ratingLevel: PropTypes.string.isRequired,
        ratingCount: PropTypes.string.isRequired,
        descriptionPartOne: PropTypes.string.isRequired,
        descriptionPartTwo: PropTypes.string.isRequired,
        filmDirector: PropTypes.string.isRequired,
        filmStarring: PropTypes.string.isRequired,
      })
  ).isRequired,
  onFilmCardClick: PropTypes.func.isRequired,
  filmsCount: PropTypes.number.isRequired,
};


export {FilmsList};
export default connect(mapStateToProps)(FilmsList);
