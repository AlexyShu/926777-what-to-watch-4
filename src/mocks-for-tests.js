export const mockFilmCard = {
  name: `The Grand Budapest`,
  genre: `Drama`,
  posterSrc: `img/the-grand-budapest-hotel-poster.jpg`,
  year: 2014,
  ratingScore: `8,9`,
  ratingLevel: `Very good`,
  ratingCount: `240 ratings`,
  descriptionPartOne: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort,
  presided over by concierge Gustave H. (Ralph Fiennes).
  Zero, a junior lobby boy, becomes Gustave's friend and protege.`,
  descriptionPartTwo: `Gustave prides himself on providing first-class service to the hotel's guests,
  including satisfying the sexual needs of the many elderly women who stay there.
  When one of Gustave's lovers dies mysteriously,
  Gustave finds himself the recipient of a priceless painting and the chief suspect in her murder.`,
  filmDirector: `Director: Wes Andreson`,
  filmStarring: `Starring: Bill Murray, Edward Norton, Jude Law, Willem Dafoe and other`,
  video: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
  time: `1h 39m`,
  reviews: [
    {
      rating: 2,
      date: `June 29, 2020`,
      author: `Alexy Shu`,
      text: `This is an awesome movie! You should watch it!`
    },
    {
      rating: 3,
      date: `June 29, 2020`,
      author: `Alexy Shu`,
      text: `This is an awesome movie! You should watch it!`
    },
    {
      rating: 9,
      date: `June 29, 2020`,
      author: `Alexy Shu`,
      text: `This is an awesome movie! You should watch it!`
    }
  ]
};

export const mockFilms = [
  {
    id: 1,
    name: `The Grand Budapest Hotel`,
    posterUrl: `img/the-grand-budapest-hotel-poster.jpg`,
    previewUrl: `img/the-grand-budapest-hotel.jpg`,
    bigPosterUrl: `img/the-grand-budapest-hotel-bg.jpg`,
    backgroundColor: `#ffffff`,
    videoUrl: `https://some-link`,
    trailerUrl: `https://some-link`,
    description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.`,
    rating: 8.9,
    votes: 240,
    director: `Wes Andreson`,
    starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`, `Saoirse Ronan`],
    runTime: `99`,
    genre: `Comedy`,
    releaseYear: 2014,
    isFavorite: false
  },
  {
    id: 2,
    name: `The Grand Budapest Hotel`,
    posterUrl: `img/the-grand-budapest-hotel-poster.jpg`,
    previewUrl: `img/the-grand-budapest-hotel.jpg`,
    bigPosterUrl: `img/the-grand-budapest-hotel-bg.jpg`,
    backgroundColor: `#ffffff`,
    videoUrl: `https://some-link`,
    trailerUrl: `https://some-link`,
    description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.`,
    rating: 8.9,
    votes: 240,
    director: `Wes Andreson`,
    starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`, `Saoirse Ronan`],
    runTime: `99`,
    genre: `Drama`,
    releaseYear: 2014,
    isFavorite: false
  },
  {
    id: 3,
    name: `The Grand Budapest Hotel`,
    posterUrl: `img/the-grand-budapest-hotel-poster.jpg`,
    previewUrl: `img/the-grand-budapest-hotel.jpg`,
    bigPosterUrl: `img/the-grand-budapest-hotel-bg.jpg`,
    backgroundColor: `#ffffff`,
    videoUrl: `https://some-link`,
    trailerUrl: `https://some-link`,
    description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.`,
    rating: 8.9,
    votes: 240,
    director: `Wes Andreson`,
    starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`, `Saoirse Ronan`],
    runTime: `99`,
    genre: `Action`,
    releaseYear: 2014,
    isFavorite: false
  }
];

export const mokcFunction = () => {};

export const MOCK_FILMS_COUNT = 8;

export const MOCK_ACTIVE_FILTER = `All genres`;

export const mockBool = false;
