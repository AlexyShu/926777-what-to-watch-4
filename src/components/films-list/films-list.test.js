import React from "react";
import renderer from "react-test-renderer";
import {FilmsList} from "./films-list.jsx";
import {mockFilms, mokcFunction, MOCK_FILMS_COUNT, mockFilmCard, mockString} from "../../mocks-for-tests.js";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import Namespace from "../../reducer/namespace.js";
import {MemoryRouter} from "react-router-dom";

const mockStore = configureStore([]);

describe(`Render correct FilmsList`, () => {
  it(`Render FilmsList`, () => {
    const store = mockStore({
      [Namespace.DATA]: {
        films: mockFilms,
        promoFilm: mockFilmCard
      },
      [Namespace.STATE]: {
        filmsCount: MOCK_FILMS_COUNT
      },
      [Namespace.USER]: {
        authorizationStatus: mockString
      }
    });
    const tree = renderer
    .create(
        <Provider store={store}>
          <MemoryRouter>
            <FilmsList
              films = {mockFilms}
              filmsCount = {MOCK_FILMS_COUNT}
              activeItem = {MOCK_FILMS_COUNT}
              handleChange = {mokcFunction}
              filteredFilms = {mockFilms}
            />
          </MemoryRouter>
        </Provider>)
    .toJSON();

    expect(tree).toMatchSnapshot();
  });
});


