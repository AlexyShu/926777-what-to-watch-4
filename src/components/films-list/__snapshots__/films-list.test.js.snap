// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render correct FilmsList Render FilmsList 1`] = `
<div
  className="catalog__movies-list"
>
  <a
    className="small-movie-card catalog__movies-card small-movie-card__link"
    href="/film-page/1"
    onClick={[Function]}
  >
    <article
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <div
        className="small-movie-card__image"
      >
        <img
          alt="The Grand Budapest Hotel"
          height="175"
          src="img/the-grand-budapest-hotel-poster.jpg"
          width="280"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="#"
        >
          The Grand Budapest Hotel
        </a>
      </h3>
    </article>
  </a>
  <a
    className="small-movie-card catalog__movies-card small-movie-card__link"
    href="/film-page/2"
    onClick={[Function]}
  >
    <article
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <div
        className="small-movie-card__image"
      >
        <img
          alt="The Grand Budapest Hotel"
          height="175"
          src="img/the-grand-budapest-hotel-poster.jpg"
          width="280"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="#"
        >
          The Grand Budapest Hotel
        </a>
      </h3>
    </article>
  </a>
  <a
    className="small-movie-card catalog__movies-card small-movie-card__link"
    href="/film-page/3"
    onClick={[Function]}
  >
    <article
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <div
        className="small-movie-card__image"
      >
        <img
          alt="The Grand Budapest Hotel"
          height="175"
          src="img/the-grand-budapest-hotel-poster.jpg"
          width="280"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="#"
        >
          The Grand Budapest Hotel
        </a>
      </h3>
    </article>
  </a>
</div>
`;
